dash/include/dash/algorithm/Accumulate.h

```diff
@@ -10,9 +9,42 @@
 
 namespace dash {
 
+namespace internal {
+
+  template<typename ValueType>
+  struct local_result {
+    ValueType value{};
+    bool      valid = false;
+  };
+
+  template<typename ValueType, typename F>
+  void accumulate_custom_fn(
+    const void *invec,
+          void *inoutvec,
+          size_t,
+          void *userdata)
+  {
+    using local_result_t = struct local_result<ValueType>;
+    const local_result_t* in    = static_cast<const local_result_t*>(invec);
+          local_result_t* inout = static_cast<local_result_t*>(inoutvec);
+    F& fn = *static_cast<F*>(userdata);
+    if (in->valid) {
+      if (inout->valid) {
+        inout->value = fn(in->value, inout->value);
+      } else {
+        inout->value = in->value;
+        inout->valid  = true;
+      }
+    }
+  }
+} // namespace internal
+
+
 /**
- * Accumulate values in range \c [first, last) as the sum of all values
- * in the range.
+ * Accumulate values in each process' range \c [first, last) using the
+ * given binary reduce function \c op, which should be a commutative operation.
```

Comment:
I don't understand why MPI says it _should_ be a commutative operation. If it isn't we have no guarantee to get the correct result (it's not a left fold anymore in our case). Even if MPI somehow does the right thing, we still apply `binary_op(result, init)` at the end.

Reply:
That has nothing to do with MPI. We cannot guarantee the correct *global* order (as you said) so the operation has to be commutative. MPI allows for non-commutative operations but we cannot use them.

Reply:
Okay, so it should be "which must be commutative" then.

Reply:
This is not sufficient, as the function has to be linear. Consider e.g. `(a,b) : a^2+b^2`

Reply:
So the binop needs to be commutative and associative: it might be a good idea to rename the algorithm to `dash::reduce`. This not only prevents confusion about the order (like with `dash::for_each`): if we additionally require `init` to be the neutral element instead of an initial value, we could get rid of a lot of "empty node" code.

Reply:
I extended the documentation a bit and included that the function has to be linear along with a description of the reduction process. 

I am also not happy with the way we copied the semantics of `std::accumulate`. First, we cannot guarantee a left fold (without significant performance issues) so the name of `dash::accumulate` might be confusing. Second, the `init` argument has to be the same on all units, otherwise the result is bogus. 

I'm happy to rename `dash::accumulate` to `dash::reduce` and change `init` to be a neutral element if there is consent to do so.